### PR TITLE
Cleaning and improvements to run Summer22/23 productions

### PR DESCRIPTION
**Pull request by BlancoFS**
Several improvements to run Summer22/23 campaigns:

- Add ttHMVA filler
- Add Lepton Tag and Probe modules
- Add module to make the Fake phase space selection
- Add several configuration files in Data to clean the Step_cfg.py file using `RPLME_CMSSW`
- Add unprescaled Run 3 HLT paths with lower thresholds
- Add data and some samples for Summer23(BPix) and Run2023(BPix)
- Add latest agreed Lepton WP ([link](https://indico.cern.ch/event/1451226/#179-lepton-wp-for-run-3))
- Add `fastparquet` in the installation requirements for the `addTnPLepton` module

### CODE REVIEW

**Comment by BlancoFS** [comment]
Please @NTrevisani @mlizzo, take a look

File: mkShapesRDF/processor/framework/Steps_cfg.py [hunk 1]
```diff
@@ -724,61 +624,15 @@
...
-        "module": "jmeCalculator()",
-    },
-    "JES_modules_Summer22EE": {
+    "JES_modules": {
         "isChain" : False,
         "do4MC"   : True,
         "do4Data" : False,
         "import"  : "mkShapesRDF.processor.modules.JMECalculator",
-        "declare" : 'jmeCalculator = lambda : JMECalculator("RPLME_FW/processor/data/jsonpog-integration/POG/JME/2022_Summer22EE/jet_jerc.json.gz", "Summer22EE_22Sep2023_V2_MC", "Summer22EE_22Sep2023_JRV1_MC", "RPLME_FW/processor/data/jsonpog-integration/POG/JME/jer_smear.json.gz", jet_object="AK4PFPuppi", jes_unc=["Total"], \
+        "declare" : 'jmeCalculator = lambda : JMECalculator("RPLME_CMSSW", jet_object="AK4PFPuppi", jes_unc=["Total"], \
```
Comment (mlizzo) [inline mkShapesRDF/processor/framework/Steps_cfg.py:632]: Hi @BlancoFS , according to what was added [here](https://github.com/latinos/mkShapesRDF/pull/49/files#diff-24cb161941ae0ff2ceaf7293ebfec5428715ff3e204de1e9be15dbf245605fc1R13), shouldn't the JMECalculator constructor have an `year` argument as well? And if found, the corresponding dictionary would be read from the new `JetMaker_cfg.py` configuration file - I like this idea by the way, but please tell me if I got this right

File: mkShapesRDF/processor/modules/JMECalculator.py [hunk 2]
```diff
@@ -66,6 +60,22 @@ def __init__(
         self.store_nominal = store_nominal
         self.store_variations = store_variations
 
+        self.base_path = ""
```
Comment (mlizzo) [inline mkShapesRDF/processor/modules/JMECalculator.py:63]: Why is this needed? Couldn't we use the `RPLME_FW` variable?

File: mkShapesRDF/processor/modules/JetSelMask.py [hunk 3]
```diff
@@ -12,10 +14,17 @@ def __init__(self, jetId, puJetId, minPt, maxEta, UL2016fix=False, pathToJson=""
         self.doMask = True
         self.UL2016fix = UL2016fix
         self.eventMask = eventMask
-        if pathToJson!="":
+
+        self.base_path = ""
```
Comment (mlizzo) [inline mkShapesRDF/processor/modules/JetSelMask.py:18]: Same as [above](https://github.com/latinos/mkShapesRDF/pull/49/files#r1867726292)

File: mkShapesRDF/processor/modules/LumiMask.py [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+from mkShapesRDF.processor.data.Lumi_cfg import LumiCfg
+import json
+import os
 
 class LumiMask(Module):
-    def __init__(self, jsonFile):
+    def __init__(self, year):
         super().__init__("LumiMask")
-        self.jsonFile = jsonFile
+        self.base_path = ""
```
Comment (mlizzo) [inline mkShapesRDF/processor/modules/LumiMask.py:9]: Same as [above](https://github.com/latinos/mkShapesRDF/pull/49/files#r1867726292)

File: mkShapesRDF/processor/data/JetMaker_cfg.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+####################### Jet Veto/JEC/JER/JES CFG ##################################
+
+JetMakerCfg = {
+    'Full2022v12': {
+        "vetomap": "/processor/data/jsonpog-integration/POG/JME/2022_Summer22/jetvetomaps.json.gz",
```
Comment (mlizzo) [inline mkShapesRDF/processor/data/JetMaker_cfg.py:5]: See comment [below](https://github.com/latinos/mkShapesRDF/pull/49/files#r1867726292) about adding `RPLME_FW` variable to find the absolute path

File: mkShapesRDF/processor/data/Lumi_cfg.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+####################### PU Weight CFG ##################################
+
+LumiCfg = {
+    'Full2022v12': "/processor/data/certification/Cert_Collisions2022_355100_362760_Golden.json",
```
Comment (mlizzo) [inline mkShapesRDF/processor/data/Lumi_cfg.py:4]: Same comment as [here](https://github.com/latinos/mkShapesRDF/pull/49/files#r1867726292)

**Comment by mlizzo** [comment]
Would it be possible to address this [issue](https://github.com/latinos/mkShapesRDF/issues/47) as well? It's a minor modification to fix a bug

File: mkShapesRDF/processor/modules/LeptonFiller_ttHMVA.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            df = df.Define("Electron_miniRelIsoNeutral", "Electron_miniPFRelIso_all - Electron_miniPFRelIso_chg")
+            df = df.Define("Electron_jetPtRatio", "getJetPtRatio(Electron_jetRelIso)")
+            df = df.Define("Electron_jetBTagDeepFlavB", "ROOT::VecOps::Take(Jet_btagDeepFlavB,Electron_jetIdx,float(0.0))")
+            df = df.Define("Electron_log_dxy", "ROOT::VecOps::log(abs(Electron_dxy))")
+            df = df.Define("Electron_log_dz", "ROOT::VecOps::log(abs(Electron_dz))")
+
+        df = df.Define(
+            "Muon_tthMVA",
+            "evaluateTTH_muon(Muon_pt, Muon_eta, Muon_pfRelIso03_all, Muon_miniPFRelIso_chg, Muon_miniRelIsoNeutral, Muon_jetNDauCharged, Muon_jetPtRelv2, Muon_jetBTagDeepFlavB, Muon_jetPtRatio, Muon_sip3d, Muon_log_dxy, Muon_log_dz, Muon_segmentComp)"
+        )
```
Comment (mlizzo) [inline mkShapesRDF/processor/modules/LeptonFiller_ttHMVA.py:201]: This is really very inefficient, as every variable and weights file are loaded on an event-basis. I think we should decouple the two things, meaning that first we have to load variables/weights - just once - and then we call a separate function to evaluate the ttHMVA score event-by-event. It should be something like [this](https://github.com/latinos/PlotsConfigurationsRun3/blob/main/FakeRate/macros/ttH_MVA_reader_class.cc)

File: mkShapesRDF/processor/modules/LeptonFiller_ttHMVA.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        df = df.Define(
+            "Electron_tthMVA",
+            "evaluateTTH_electron(Electron_pt,Electron_eta,Electron_pfRelIso03_all,Electron_miniPFRelIso_chg,Electron_miniRelIsoNeutral,Electron_jetNDauCharged,Electron_jetPtRelv2,Electron_jetBTagDeepFlavB,Electron_jetPtRatio,Electron_sip3d,Electron_log_dxy,Electron_log_dz,Electron_mvaIso)"
+        )
```
Comment (mlizzo) [inline mkShapesRDF/processor/modules/LeptonFiller_ttHMVA.py:205]: Same as of the muon_ttHMVA

File: mkShapesRDF/processor/framework/processor.py [hunk 9]
```diff
@@ -345,6 +345,7 @@ def sciNot(value):
         """)
         
         self.fPy = self.fPy.replace("RPLME_FW", frameworkPath)
+        self.fPy = self.fPy.replace("RPLME_CMSSW", Productions[self.prodName]["cmssw"])
```
Comment (mlizzo) [inline mkShapesRDF/processor/framework/processor.py:348]: I don't like this name at all, why "cmssw"? I know it's just a name, but I don't see why we have to call it this way, considering that it has nothing to do with cmssw - it's just the production name iiuc

File: mkShapesRDF/processor/framework/samples/Summer23_130x_nAODv12.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+} 
+
+Samples['VBFHToWWTo2L2Nu_M125'] = {
+    'nanoAOD' :	'/VBFHto2Wto2L2Nu_M-125_TuneCP5_13p6TeV_powheg-jhugen752-pythia8/Run3Summer23NanoAODv12-130X_mcRun3_2023_realistic_v15-v2/NANOAODSIM'
+}
+
+##### ggWW !!!! 
+
+#Samples['GluGlutoContintoWWtoENuENu'] = {
+#    'nanoAOD' :''
```
Comment (mlizzo) [inline mkShapesRDF/processor/framework/samples/Summer23_130x_nAODv12.py:16]: FYI: [ggWW Run3Summer23BPixNanoAODv12 samples](https://cmsweb.cern.ch/das/request?view=list&limit=50&instance=prod%2Fglobal&input=dataset%3D%2FGluGlutoContintoWW*_TuneCP5_13p6TeV_mcfm701-pythia8%2F*Run3Summer23NanoAODv12*%2FNANOAODSIM)

File: mkShapesRDF/processor/framework/samples/Summer23BPix_130x_nAODv12.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+} 
+
+Samples['VBFHToWWTo2L2Nu_M125'] = {
+    'nanoAOD' :	'/VBFHto2Wto2L2Nu_M-125_TuneCP5_13p6TeV_powheg-jhugen752-pythia8/Run3Summer23BPixNanoAODv12-130X_mcRun3_2023_realistic_postBPix_v6-v2/NANOAODSIM'
+}
+
+##### ggWW !!!! 
+
+#Samples['GluGlutoContintoWWtoENuENu'] = {
+#    'nanoAOD' :''
```
Comment (mlizzo) [inline mkShapesRDF/processor/framework/samples/Summer23BPix_130x_nAODv12.py:16]: FYI: [ggWW Run3Summer23BPixNanoAODv12 samples](https://cmsweb.cern.ch/das/request?view=list&limit=50&instance=prod%2Fglobal&input=dataset%3D%2FGluGlutoContintoWW*_TuneCP5_13p6TeV_mcfm701-pythia8%2F*Run3Summer23BPixNanoAODv12*%2FNANOAODSIM)

File: mkShapesRDF/processor/framework/Steps_cfg.py (hunk 1 above)
Comment (BlancoFS) [reply mkShapesRDF/processor/framework/Steps_cfg.py:632]: Hi @mlizzo, sorry for the late response and thank you so much for all the comments! I will try to go over them later. I will just try to explain the logic of this one. 

The JMECalculator does have a `year` argument. The point is that, as in `LatinoAnalysis`, the `RPLME_CMSSW` key is the `year` taken from Productions_cfg.py, substituted when creating the jobs and passed to the JMECalculator constructor as an argument. Then, it takes the rests of the configurations from `JetMaker_cfg.py`. 

In this way, we have general steps that can be used for all the campaigns. There is no need to define one task per year in the Steps_cfg.py file. It's just a way of having a more simple and clear Steps_cfg.py.

Of course, the `cmssw` key from Productions_cfg.py ([example](https://github.com/latinos/mkShapesRDF/blob/60de689e513dc726295fde020bfda124e0b23ce5/mkShapesRDF/processor/framework/Productions_cfg.py#L123)) should match the one in JetMaker_cfg.py ([same example](https://github.com/latinos/mkShapesRDF/blob/60de689e513dc726295fde020bfda124e0b23ce5/mkShapesRDF/processor/data/JetMaker_cfg.py#L4)).

File: mkShapesRDF/processor/modules/JMECalculator.py (hunk 2 above)
Comment (BlancoFS) [reply mkShapesRDF/processor/modules/JMECalculator.py:63]: This is needed because the `RPLME_*` variables are substituted when creating the `script.py` file with the information from `Steps_cfg.py`. However, the `processor.py` does not modify the content of the modules

File: mkShapesRDF/processor/framework/Productions_cfg.py [hunk 12]
```diff
@@ -50,9 +50,23 @@
...
         "jsonFile": "../data/certification/Cert_Collisions2022_355100_362760_Golden.json",
         "samples": "../framework/samples/Run2022EE_Prompt_nAODv12.py",
-        "cmssw": "Full2022v12",
+        "cmssw": "Full2022EEv12",
         "year": "2022",
     },
+    "Run2023_Prompt_nAODv12_Full2023v12": {
+        "isData": True,
+        "jsonFile": "../data/certification/Cert_Collisions2023_366442_370790_Golden.json",
+        "samples": "../framework/samples/Run2023_Prompt_nAODv12.py",
```
[the remark this answers is not in the document]
Comment (BlancoFS) [reply mkShapesRDF/processor/framework/Productions_cfg.py:59]: Ah! I didn't noticed that the information from the Lumi mask was not doubled, but even tripled in the previous iteration. I created the `Lumi_cfg.py` file to avoid defining the relative paths from `Steps_cfg.py` but I didn't noticed that the information was also writen in the `Productions_cfg.py` file.


I think the relative paths are not a real problem since the absolute path is created in the module. I will try to find a way of not duplicating the information and use the path defined in `Productions_cfg.py`

File: mkShapesRDF/processor/framework/processor.py (hunk 9 above)
Comment (BlancoFS) [reply mkShapesRDF/processor/framework/processor.py:348]: It's completely true. Honestly, I just used the same name that was used in LatinoAnalysis but of course might be changed

File: mkShapesRDF/processor/modules/JMECalculator.py (hunk 2 above)
Comment (mlizzo) [reply mkShapesRDF/processor/modules/JMECalculator.py]: But couldn't we simply add the `RPLME_FW` header in `Steps_cfg.py` and `Productions_cfg.py` configuration files, whenever extra inputs are called? Wouldn't this be substituted at run time by the `processor.py` module?

**Comment by mlizzo** [comment]
Hi @BlancoFS , thank you very much for promptly updating files according to our discussion and for the huge work you did, I really appreciate it. I just left another comment related to the usage of the `RPLME_FW` variable. Other than that, would you mind to quickly fix this [issue](url) as well? It's just a matter of changing one line, but it's important to tackle if we want to move on with Run3 production. Lastly, have you tested this PR? There are a lot of modifications, and before merging them I would like to make sure that they don't break the code. In the (near) future I think we should set up a github pipeline to automatically check that everything runs as expected, but since we don't have this yet, we are bounded to do this manually for the moment being. 

 I don't have other feedbacks to provide, @NTrevisani @dittmer please have a look as well.
 
 Thanks again Sergio, that's great!

File: mkShapesRDF/processor/modules/JMECalculator.py (hunk 2 above)
Comment (BlancoFS) [reply mkShapesRDF/processor/modules/JMECalculator.py]: I have changed the logic to make more or less what you say, so now the `RPLME_FW` key is used in the Steps_cfg.py file as arguments for JMECalculator. Here you can see an example:

https://github.com/latinos/mkShapesRDF/blob/5eef26cc64aa690749442f801507a8fc152f7be2/mkShapesRDF/processor/framework/Steps_cfg.py#L632

I think this is what you mean, right?

**Comment by BlancoFS** [comment]
Sorry @mlizzo but it seems like the link is broken. Which is the issue you mean?

I have tested that the postprocessing step works correctly for Data and MC. However, I haven't tested the content of the output ntuples, since this PR only modifies how the code is structured and not the content of any module.

**Comment by mlizzo** [comment]
Sorry for that, the link is now fixed

File: mkShapesRDF/processor/framework/Steps_cfg.py (hunk 1 above)
Comment (mlizzo) [reply mkShapesRDF/processor/framework/Steps_cfg.py]: Sorry, I think I wasn't very clear indeed. If I understand correctly, you have added `JetMaker_cfg.py` as a new steering file, which is basically an auxiliary configuration to setup the JME module. What I meant was to replace relative paths with absolute ones [here](https://github.com/latinos/mkShapesRDF/blob/5eef26cc64aa690749442f801507a8fc152f7be2/mkShapesRDF/processor/data/JetMaker_cfg.py#L12-L19), by simply adding `RPLME_FW` as header. If you do so, you should be able to remove the `base_path` argument, as it would be unnecessary in this way - being implicitly included in the `JetMaker_cfg.py` file. It's silly, but at least you don't have to carry an extra argument just for that

**Comment by BlancoFS** [comment]
Aaah it's should be already solved here:

https://github.com/latinos/mkShapesRDF/blob/5eef26cc64aa690749442f801507a8fc152f7be2/mkShapesRDF/processor/modules/LeptonSF.py#L455

**Comment by mlizzo** [comment]
Everything looks good to me, don't have further comments from my side. @NTrevisani @dittmer I'm gonna merge this PR if you don't raise any objections

**Comment by NTrevisani** [comment]
Thank you @BlancoFS for preparing the pull request and @mlizzo for the thorough review. I have no further comments and I agree with merging the pull request.
